### PR TITLE
[shopsys] upgrade easy coding standards

### PR DESCRIPTION
**Pull request by grossmannmartin**
| Q             | A
| ------------- | ---
|Description, reason for the PR| Upgraded easy-coding-standard package to 12.2. Configuration changed to new format. Packages now do not run ecs separately as it does not make sense (it's impossible to commit to individual packages without monorepo). Cleanup of skip rules. Paths are now part of configuration.
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://mg-upgrade-ecs.odin.shopsys.cloud
  - https://cz.mg-upgrade-ecs.odin.shopsys.cloud

### CODE REVIEW

File: ecs.php [hunk 1]
```diff
@@ -3,44 +3,170 @@
...
- */
-return static function (ECSConfig $ecsConfig): void {
-    $ecsConfig->rule(PhpdocToPropertyTypeFixer::class);
-    $ecsConfig->skip(
+return ECSConfig::configure()
+    ->withPaths([
+        __DIR__ . '/project-base/app/app',
+        __DIR__ . '/project-base/app/src',
+        __DIR__ . '/project-base/app/tests',
+        __DIR__ . '/packages/article-feed-luigis-box/src',
```
Comment (vitek-rostislav) [inline ecs.php]: `packages/*/src` is not an option anymore, right? :disappointed: I know the upgrading is hell without the upgrading instructions :frowning_face:

File: ecs.php [hunk 2]
```diff
@@ -3,44 +3,170 @@
...
+                'Shopsys\ProductFeed\ZboziBundle\Model',
+                'Shopsys\S3Bridge',
+            ],
+        ],
+    )
+    ->withConfiguredRule(CyclomaticComplexitySniff::class, [
+        'absoluteComplexity' => CyclomaticComplexitySniffSetting::DEFAULT_ABSOLUTE_COMPLEXITY,
+    ])
+    ->withSkip(array_merge_recursive(
+        include __DIR__ . '/project-base/app/ecs-skip-rules.php',
```
Comment (vitek-rostislav) [inline ecs.php:86]: smart solution :bulb: :ok_hand:

File: packages/coding-standards/ecs.php [hunk 3]
```diff
@@ -140,180 +131,174 @@
...
+        ForbiddenDoctrineInheritanceSniff::class,
+        ForbiddenDoctrineDefaultValueSniff::class,
+        // method arguments and variables should be $camelCase
+        ValidVariableNameSniff::class,
+        // add all @param, @return and @var annotations, in FQN
+        MissingParamAnnotationsFixer::class,
+        MissingReturnAnnotationFixer::class,
+        OrderedParamAnnotationsFixer::class,
+        FullyQualifiedClassNameInAnnotationSniff::class,
+        // Custom Shopsys standards
```
Comment (vitek-rostislav) [inline packages/coding-standards/ecs.php]: I know it has been there already, however, I am not sure whether I understand this comment and I feel like I want to get rid of it :smile:

Comment (vitek-rostislav) [inline packages/coding-standards/ecs.php]: I know it has been there already, however, I am not sure whether I understand this comment and I feel like I want to get rid of it :smile:

File: packages/coding-standards/ecs.php [hunk 4]
```diff
@@ -329,63 +314,43 @@
...
-    ]);
-
-    $ecsConfig->skip([
-        __DIR__ . '/tests/Unit/**/wrong/*',
-        __DIR__ . '/tests/Unit/**/Wrong/*',
-        __DIR__ . '/tests/Unit/**/correct/*',
-        __DIR__ . '/tests/Unit/**/Correct/*',
-        __DIR__ . '/tests/Unit/**/fixed/*',
+    ])
+    ->withSkip([
```
Comment (vitek-rostislav) [inline packages/coding-standards/ecs.php:336]: could we please discuss this briefly? Is not there a problem that project-base includes the config from `shopsys/coding-standards` and defines its own `withSkip`?

File: project-base/app/ecs-skip-rules.php [hunk 5]
```diff
@@ -25,93 +25,45 @@
         __DIR__ . '/tests/App/Functional/EntityExtension/Model/*',
     ],
     FunctionLengthSniff::class => [
-        __DIR__ . '/src/Migrations/Version20190801103940.php',
```
Comment (vitek-rostislav) [inline project-base/app/ecs-skip-rules.php]: the `ecs` tool is now reporting the useless skips or was this a manual work? :open_mouth:

File: UPGRADE-15.0.md [hunk 6]
```diff
@@ -861,6 +861,14 @@ Follow the instructions in relevant sections, e.g. `shopsys/coding-standards` or
 -   ParameterColorValueDataFixture is added to handle assigning hex values to color parameters
 -   see #project-base-diff to update your project
 
+#### update easy-coding-standard to version 12.2 ([#3192](https://github.com/shopsys/shopsys/pull/3192))
+
+-   update configuration file to new version
```
Comment (vitek-rostislav) [inline UPGRADE-15.0.md:979]: once again, it is very sad that we have no upgrade guide to link to. I am afraid people will be quite sad but I do not feel like writing upgrade notes for another package

**Review by vitek-rostislav — APPROVED** [review]

File: packages/coding-standards/ecs.php (hunk 4 above)
Comment (grossmannmartin) [reply packages/coding-standards/ecs.php:336]: skips are not a problem in general, but the problem is when some package defines skip and another has the same rule skipped with other files.
e.g., 

```
// project-base/ecs.php

->withSkip([
    FunctionLength::class => [
        __DIR__ . '/file1',
        __DIR__ . '/file2',
    ],
    OneSniff::class,
]);

// monorepo ecs.php

->withSets([
    __DIR__ . '/project-base/ecs.php',
])
->withSkip([
    FunctionLength::class => [
        __DIR__ . '/file3',
        __DIR__ . '/file4',
    ],
    TwoSniff::class,
]);

```

Both OneSniff and TwoSniff are skipped, but the list of files for FunctionLength is not merged, but overridden.

File: project-base/app/ecs-skip-rules.php (hunk 5 above)
Comment (grossmannmartin) [reply project-base/app/ecs-skip-rules.php]: it was manual work... sadly ecs does not report useless skips

File: UPGRADE-15.0.md (hunk 6 above)
Comment (grossmannmartin) [reply UPGRADE-15.0.md:979]: Yes... I hope the project-base diff give at least a general idea about necessary changes

File: ecs.php (hunk 1 above)
Comment (grossmannmartin) [reply ecs.php]: yes, it's not possible to use an asterisk... 

we could create the list of paths automatically with DirectoryIterator and file_exists, though 🤔

File: packages/coding-standards/ecs.php (hunk 3 above)
Comment (grossmannmartin) [reply packages/coding-standards/ecs.php]: removed unnecessary comments in bbaa1a5686

Comment (grossmannmartin) [reply packages/coding-standards/ecs.php]: same as above

**Review by pk16011990 — APPROVED** [review]

**Pull request merged**